### ISSUE TITLE
Error: NDArray is disposed.

### ISSUE THREAD

**Issue by rwieruch** [PAIR-code/deeplearnjs#440]
Recently I have built a NN with deeplearn.js. It works without problems in development mode on my machine. But after building the project and deploying it to production, I get `Error: NDArray is disposed.` Sometimes the error shows only up after a while and sometimes early in the beginning.

GitHub: https://github.com/javascript-machine-learning/color-accessibility-neural-network-deeplearnjs
Live: http://color-accessibility-neural-network-deeplearnjs.bri.im/

<img width="1015" alt="screen shot 2017-12-14 at 11 47 33" src="https://user-images.githubusercontent.com/2479967/33974667-9fed6272-e0c4-11e7-8e71-85a5e9940214.png">

How can development and production be any different? The cost doesn't update in production. Is it on my side because of the build step? Or is the error somewhere in my view layer because of `requestAnimationFrame`? Any hint to solve it would be awesome.

**Comment by anti-destiny** [comment]
I'm having the same problem. I'm starting up with the demo "complementary color prediction" and the error "NDArray is disposed" appeared in my console. 
My build command is "browserify -d .\complementary-color-prediction.ts -p [tsify] -o ./bundle.js" and my node version is v8.9.0, my deeplearn version is 0.3.14.

**Comment by dsmilkov** [comment]
It is strange that development mode works, but production doesn't. That's something we should definitely investigate.

In addition, since `0.3.14`, we have global tracking of NDArrays, which is a backwards incompatible change. This change allows us to catch all memory leaks. Apps that used to have memory leaks will now face the problem of `NDArray` being disposed.

Two options:

1) Quick solution to unblock you: downgrade to `0.3.13`. Our demos are still using deeplearn `0.3.13` as a peerDependency. We are actively working on migrating our demos to `0.3.14`. Once we migrate, you can fork our demos.
2) Longer-term solution: try to fix usages of `math.scope`. As soon as `math.scope(() => {..})` ends, all `NDArray`s inside of it will be disposed. For example, in the repo you shared, as soon as the scope in prepareTrainingSet ends ([here](https://github.com/javascript-machine-learning/color-accessibility-neural-network-deeplearnjs/blob/master/src/neuralNetwork.js#L70)), all your training data is disposed (`inputArray` and `targetArray`). Since you go over your training data many times over, remove the math.scope in `prepareTrainingSet()`. There might be other problems like that too.

Hope this helps, and apologies for the breakage!

**Comment by nsthorat** [comment]
There's also a problem with the way we do shuffled CPU providers:
https://github.com/PAIR-code/deeplearnjs/blob/master/src/data/input_provider.ts#L69

The constructor takes an NDArray[][] that is assumed to be CPU-resident. Now we upload directly to the GPU so the initial values get disposed in the outer scope. I'll prepare a fix.

**Comment by nsthorat** [comment]
Apologies about the confusion, but ignore my last comment. Daniel's comment holds - you just need to remove math.scope from the method where you create the data providers.

**Comment by rwieruch** [comment]
Thanks for your quick response!

I followed the (2) advice and removed the math.scope when preparing the trainings set. It works still on localhost, but in production there is another error showing up then:

<img width="1273" alt="screen shot 2017-12-16 at 10 25 48" src="https://user-images.githubusercontent.com/2479967/34066588-e03602d4-e24b-11e7-9cdb-c9f09fe64c3c.png">

I was running on 0.3.11, so I tried to upgrade to 0.3.13:

It works. With or without removing math.scope from the training set preparation. But I guess that's not intended, because as you said, these NDArrays should be disposed. So I upgraded to 0.3.14:

<img width="1022" alt="screen shot 2017-12-16 at 10 46 56" src="https://user-images.githubusercontent.com/2479967/34066733-8a40c050-e24e-11e7-8eac-84cf7555fd5f.png">

Again the error. But at least it happens in development and production now 👍

For now, it works with 0.3.13 on development and production (even though ). 

(1) Should I open another Issue for the error?
(2) I noticed that the performance of my NN is performing worse in 0.3.13 compared to 0.3.11. I don't have any performance measures, but it seems that it get more likely stuck in local minima (cost doesn't reduce) at an early stage.

**Comment by vuoristo** [comment]
In 0.3.14, exiting `math.scope` after `Session.train` has been called seems to cause disposal of the internal variables of Operations such as `ReduceSum.ones` and `Square.two`.

Is this the expected behavior? For my use case, I would like to keep the graph around for interleaved calls to `Session.train` and `Session.eval`. I think wrapping them into the same `math.scope` would require manual disposal of the training data I'm generating. What is the correct way to deal with the graph in this case?

Here's a code sample that demonstrates the issue:

```
class Demonstrator {
  constructor() {
    this.math = new NDArrayMathCPU();
    this.graph = new Graph();
    let g = this.graph;
    this.optimizer = new AdamOptimizer(0.0, 0.9, 0.999);

    this.inputTensor = g.placeholder('inputs', []);
    this.costTensor = g.reduceSum(this.inputTensor);

    this.session = new Session(g, this.math);
  }

  train() {
    this.math.scope((keep, track) => {
      const inputs = [];
      for (let i = 0; i < 1000; i++) {
        inputs.push(Scalar.new(i));
      }

      const shuffledInputProviderBuilder =
        new InCPUMemoryShuffledInputProviderBuilder([inputs]);
      const [inputsProvider] =
        shuffledInputProviderBuilder.getInputProviders();
      const mapping = [
        {tensor: this.inputTensor, data: inputsProvider},
      ];

      this.session.train(
        this.costTensor, mapping, inputs.length, this.optimizer,
        CostReduction.NONE);
    });
  }
}

const demo = new Demonstrator();
// On the second train call, ReduceSum internal constant 'one' will have been disposed
for (let i = 0; i < 2; i++) {
  demo.train();
}
```

**Comment by dsmilkov** [comment]
Thanks for bringing this to our attention!

A quick fix will be to install `0.3.13` instead of `0.3.14`.

In the meantime, we are working on a fix for session/optimizers/graph to work with the new global math introduced in `0.3.14`

**Comment by nsthorat** [comment]
We just released 0.3.15. In general, you should remove any explicit `NDArrayMathGPU` constructions, rather just use "dl.ENV.math" which will choose the best backend.

If you have problems let us know!

**Comment by rwieruch** [comment]
@nsthorat is there any documentation around using dl.ENV.math in favor of NDArrayMathGPU? Not sure if everyone gets how to use it as replacement for NDArrayMathGPU.

**Comment by dsmilkov** [comment]
Good q! We've deprecated `NDArrayMathGPU` constructor but there is nothing special about `dl.ENV.math`. I'm updating the old documentation in #461 to only use `dl.ENV.math`.

We introduced this global math as a transition to an upcoming simpler API where all operations are chained:

```js
a = dl.array1d([5, 3, 1]);
console.log(await a.log().sum().data()); // log(5) + log(3) + log(1)
```

All these commands are going to be executed on the global `dl.ENV.math`, which users will never need to directly reference.

We might also provide a non-chained top level api, based on feedback when we release the chained api:

```js
a = dl.array1d([5, 3, 1]);
console.log(await dl.sum(dl.log(a)).data());
```
